src/basic/alloc-util.h

```diff
@@ -242,4 +242,18 @@ static inline size_t malloc_sizeof_safe(void **xp) {
                 (char*) memdupa_suffix0(_t, strnlen(_t, (n)));          \
         })
 
+/* Free every element of the array. */
+static inline void free_many(void **p, size_t n) {
+        if (!p)
+                return;
+
+        for (size_t i = 0; i < n; i++)
+                p[i] = mfree(p[i]);
+}
+
+/* Typesafe wrapper for char** rather than void**. Unfortunately C won't implicitly cast this. */
+static inline void free_many_charp(char **c, size_t n) {
```

Comment:
Maybe create a wrapper macro for the function? Something like
```c
#define free_many(p, n) _free_many(p, n, UNIQ_T(p, UNIQ))
#define _free_many(p, n, x) ({ \
        void *x = &(**(p)); /* make sure p is a double pointer */ \
        free_many_real(&x, n); \
})
```

Reply:
I think something like this would make sense. But I think such a macro should probably generate a function. CLEANUP_ARRAY() after all wants a proper function (which it can use via the cleanup logic), it can't take a macro... And CLEANUP_ARRAY() is one primary usecase for the whole concept.

That's why I added this as a function, manually, actually, for now.

But I think we should maybe add a macro to automatically generate a function in a typesafe way, that just generates static inline type-safe versions of free_many(). However, I'd really wait until we have more users of that.